openfl-tutorials/interactive_api/Flax_CNN_CIFAR/README.md

```diff
@@ -0,0 +1,104 @@
+# Federated FLAX CIFAR-10 CNN Tutorial
+
+### 1. About dataset
+
+The CIFAR-10 dataset consists of 60000 32x32 colour images in 10 classes, with 6000 images per class. There are 50000 training images and 10000 test images.
+
+Define the below param in envoy.yaml config to shard the dataset across participants/envoy.
+- rank_worldsize
+
+### 2. About model
+
+A simple multi-layer CNN is used with XLA compiled and Auto-grad based parameter updates.
+Definition provided in the notebook.
+
+### 3. Notebook Overview
+
+1. Class `CustomTrainState` - Subclasses `flax.training.TrainState`
+    - Variable `opt_vars` to keep track of generic optimizer variables.
+    - Method `update_state` to update the OpenFL `ModelInterface` registered state with the new_state returned within the `TaskInterface` registered training loop.
+
+2. Method `create_train_state`: Creates a new `TrainState` by encapsulating model layer definitions, random model parameters, and optax optimizer state.
+
+3. Method `apply_model` (`@jax.jit` decorated function): It takes a TrainState, images, and labels as parameters. It computes and returns the gradients, loss, and accuracy. These gradients are applied to a given state in the `update_model` method (`@jax.jit` decorated function) and a new TrainState instance is returned.
+
+### 4. How to run this tutorial (without TLS and locally as a simulation):
+
+0. Pre-requisites:
```

Comment:
Can this tutorial be modified to run on CPU as well?

Reply:
Yes Patrick, just by setting this one variable **DEFAULT_DEVICE='CPU|GPU'** in `start_envoy.sh` and in workspace notebook.

**Details:**

The `jaxlib==0.3.10+cuda11.cudnn805` library version used in `requirements.txt` is by default CUDA device compatible and takes precedence during execution. If no CUDA device is present then the JAX execution falls back to CPU.

Another way we can enforce the execution on any platform is by using the environment variable `JAX_PLATFORMS='CPU|GPU|TPU'` (Updated README.md). Defined in both `start_envoy.sh` and workspace notebook.

This `JAX_PLATFORMS` environment variable need to be set prior to running `./start_envoy.sh` and notebook because the `import jax` statement allocates the GPU memory upfront so does the TensorFlow.

There are other supporting environment variables that goes along with `JAX_PLATFORMS` for the execution to work as expected without taking up GPU memory at its entirety, I am abstracting those away and user would now have to just set one variable `DEFAULT_DEVICE` in `start_envoy.sh` and workspace notebook to control the execution platform.